proof/access-control/AARCH64/ArchAccess.thy

```diff
@@ -0,0 +1,363 @@
+  sata_asid:
+    "\<lbrakk> caps ptr = Some (ArchObjectCap acap); asid \<in> acap_asid' acap \<rbrakk>
+       \<Longrightarrow> (pasObjectAbs aag (fst ptr), Control, pasASIDAbs aag asid)
+             \<in> state_asids_to_policy_aux aag caps asid_tab vrefs"
+| sata_asid_lookup:
+    "\<lbrakk> asid_tab (asid_high_bits_of asid) = Some poolptr;
+       (pdptr, ucast (asid && mask asid_low_bits), AASIDPool, a) \<in> vrefs poolptr \<rbrakk>
+       \<Longrightarrow> (pasASIDAbs aag asid, a, pasObjectAbs aag pdptr)
+             \<in> state_asids_to_policy_aux aag caps asid_tab vrefs"
+| sata_asidpool:
+    "\<lbrakk> asid_tab (asid_high_bits_of asid) = Some poolptr; asid \<noteq> 0 \<rbrakk>
+       \<Longrightarrow> (pasObjectAbs aag poolptr, AAuth ASIDPoolMapsASID, pasASIDAbs aag asid)
+             \<in> state_asids_to_policy_aux aag caps asid_tab vrefs"
+
+definition
+  "state_asids_to_policy_arch aag caps astate vrefs \<equiv>
+     state_asids_to_policy_aux aag caps (arm_asid_table astate)
+                               (vrefs :: 64 word \<Rightarrow> (64 word \<times> 64 word \<times> aa_type \<times> auth) set)"
+declare state_asids_to_policy_arch_def[simp]
+
+section \<open>Arch-specific integrity definition\<close>
+
+subsection \<open>How ASIDs can change\<close>
+
+abbreviation integrity_asids_aux :: "'a PAS \<Rightarrow> 'a set \<Rightarrow> obj_ref \<Rightarrow> asid \<Rightarrow>
+                                     (asid_high_index \<rightharpoonup> obj_ref) \<Rightarrow> (asid_high_index \<rightharpoonup> obj_ref) \<Rightarrow>
+                                     (obj_ref \<rightharpoonup> asid_pool) \<Rightarrow> (obj_ref \<rightharpoonup> asid_pool) \<Rightarrow> bool" where
+  "integrity_asids_aux aag subjects x asid atab atab' pools pools'  \<equiv>
+     (atab (asid_high_bits_of asid) \<noteq> atab' (asid_high_bits_of asid)
+      \<longrightarrow> (\<forall>x. atab' (asid_high_bits_of asid) = Some x \<longrightarrow> pasObjectAbs aag x \<in> subjects) \<and>
+          (\<forall>asid'. asid' \<noteq> 0 \<and> asid_high_bits_of asid' = asid_high_bits_of asid
+                   \<longrightarrow> pasASIDAbs aag asid' \<in> subjects)) \<and>
+     (pasObjectAbs aag x \<notin> subjects
+      \<longrightarrow> (\<forall>pool. pools x = Some pool
+                  \<longrightarrow> pools' x \<noteq> None \<and>
+                      (\<forall>pool'. pools' x = Some pool' \<longrightarrow>
+                               (pool \<noteq> pool' \<longrightarrow> (\<exists>asid. atab (asid_high_bits_of asid) = Some x)))))"
+
+definition integrity_asids ::
+  "'a PAS \<Rightarrow> 'a set \<Rightarrow> obj_ref \<Rightarrow> asid \<Rightarrow> 'y::state_ext state \<Rightarrow> 'z::state_ext state  \<Rightarrow> bool" where
+  "integrity_asids aag subjects x asid s s' \<equiv>
+   integrity_asids_aux aag subjects x asid (asid_table s) (asid_table s')
+                                           (asid_pools_of s) (asid_pools_of s')"
```

Comment:
this is interestingly the other way round to the `_2` setup we use in AInvs/Refine. There we want the abbreviation to expose the state projections, and the definition to hide the complicated bits and pieces. Here you get no benefit from the definition (need to unfold to benefit from the fact that the state projections don't change), and you don't get a real benefit from the abbreviation, because it doesn't hide the complexity ... thoughts?

Reply:
It might be worth trying out what happens if we do it the "right" way around. It might conflict with old style in the Access session, but it'd be good to see if that is true.

Reply:
I'll take a look, but refactoring would involve exposing the abbreviations. For example, `asid_pools_of s` would become `(λs. kh |> aobj_of |> asid_pool_of)` for some generic `kh` instantiated with `kheap s`